### ISSUE TITLE
Abbreviated genitive month names won't work on custom DateTimeFormat strings

### ISSUE THREAD

**Issue by javiercampos** [dotnet/corefx#2804]
As found out by @jskeet , abbreviated genitive month names don't seem to work when using the "MMM" token on custom datetime format strings

The test, written by @jskeet (I'm just copying here from https://github.com/nodatime/nodatime/issues/377):

    using System;
    using System.Globalization;
    using System.Linq;

    class Test
    {
        static void Main()
        {
            var culture = (CultureInfo) CultureInfo.InvariantCulture.Clone();
    
            culture.DateTimeFormat.AbbreviatedMonthGenitiveNames
                = culture.DateTimeFormat.AbbreviatedMonthNames
                         .Select(x => "Gen" + x)
                         .ToArray();
            culture.DateTimeFormat.MonthGenitiveNames
                = culture.DateTimeFormat.MonthNames
                         .Select(x => "Gen" + x)
                         .ToArray();

            Console.WriteLine(new DateTime(1976, 6, 19).ToString("d MMMM yy", culture));
            Console.WriteLine(new DateTime(1976, 6, 19).ToString("MMMM yy", culture));
            Console.WriteLine(new DateTime(1976, 6, 19).ToString("d MMM yy", culture));
            Console.WriteLine(new DateTime(1976, 6, 19).ToString("MMM yy", culture));
        }
    }

Should output:

    19 GenJune 76
    June 76
    19 GenJun 76
    Jun 76

But outputs:

    19 GenJune 76
    June 76
    19 Jun 76
    Jun 76

Disregarding the abbreviated genitives (but not the longer one).

I haven't found the DateTimeFormat files on GitHub yet (just the tests, which don't cover this), so I can't issue a request or test it, but on the reference source: http://referencesource.microsoft.com/#mscorlib/system/globalization/datetimeformat.cs,578

We got this:

    if ((dtfi.FormatFlags & DateTimeFormatFlags.UseGenitiveMonth) != 0 
                  && tokenLen >= 4) {
            result.Append(
                 dtfi.internalGetMonthName(
                        month, 
                        IsUseGenitiveForm(format, i, tokenLen, 'd')? MonthNameStyles.Genitive : MonthNameStyles.Regular, 
                        false));

Which doesn't take MMM into account (because of the `tokenLen >= 4` conditional). If I'm not mistaken, it should look like this:

    if ((dtfi.FormatFlags & DateTimeFormatFlags.UseGenitiveMonth) != 0 
                  && tokenLen >= 3 /* Work for MMM and MMMM */ ) {
            result.Append(
                 dtfi.internalGetMonthName(
                        month, 
                        IsUseGenitiveForm(format, i, tokenLen, 'd')? MonthNameStyles.Genitive : MonthNameStyles.Regular, 
                        tokenLen == 3 /* Abbreviated if MMM */));

**Comment by weshaggard** [comment]
cc @tarekgh

**Comment by tarekgh** [comment]
The AbbreviatedMonthGenitiveNames is not really used at all in the formatting and we always use AbbreviatedMonthNames. if you look at the doc:
https://msdn.microsoft.com/en-us/library/system.globalization.datetimeformatinfo(v=vs.110).aspx
it states:
> "MMM" (custom format specifier)
> AbbreviatedMonthNames, to include the abbreviated month name in the result string. 
> "MMMM" (custom format specifier)
> MonthNames or MonthGenitiveNames, to include the full month name in the result string.
 
Note that the doc didn't mentioning the usage of AbbreviatedMonthGenitiveNames with "MMM" and always use AbbreviatedMonthNames while it mention MonthGenitiveNames with "MMMM"

Also the doc:
https://msdn.microsoft.com/en-us/library/system.globalization.datetimeformatinfo.abbreviatedmonthgenitivenames(v=vs.110).aspx

States the following in the Remarks section:

> If you set the AbbreviatedMonthGenitiveNames property, you must also set the AbbreviatedMonthNames property.

so it looks this is done intentionally but I don't know the reason behind that. I'll try to find out the history of that. anyway if this need to change in the desktop will need to be quirked as it is a breaking change.

**Comment by javiercampos** [comment]
The doc at https://msdn.microsoft.com/en-us/library/system.globalization.datetimeformatinfo.abbreviatedmonthgenitivenames(v=vs.110).aspx actually states (bolds are mine):

> In some languages, a month name that is part of a date appears in the genitive case. For example, a date in the ru-RU or Russian (Russia) culture consists of the day number and the genitive month name, such as 1 Января (1 January). For these cultures, **if a custom format string includes the "MMM" format specifier, the DateTime.ToString or ToString method includes the appropriate member of the AbbreviatedMonthGenitiveNames array in place of the "MMM" in the result string.**

So according to "that" doc it should do it. Don't know why it's not stated in the `DateTimeFormatInfo` one.

**Comment by tarekgh** [comment]
@javiercampos yes that is because it always assume AbbreviatedMonthNames is set when  AbbreviatedMonthGenitiveNames is,

> If you set the AbbreviatedMonthGenitiveNames property, you must also set the AbbreviatedMonthNames property.

Also that is what is happening in the code too. 
As I mentioned I don't know the historical reason for that but trying to find out.

**Comment by tarekgh** [comment]
I couldn't know the historical reason for current behavior but I believe we need to fix it. the fix would not be only changing the length check of (MMM) but we'll need to look at different areas to make sure AbbreviatedMonthGenitiveNames is used when it is really needed.

Changing this on the desktop will require quirking the behavior to keep the app compat safe.
